### ISSUE TITLE
Configure misdetects setlocale() features if run twice

### ISSUE THREAD

**Issue by mauke** [Perl/perl5#21406]
**Description**
If `config.sh` exists (which it does after running `Configure` at least once), any subsequent run of `Configure` will misdetect how `setlocale()` works. In particular, it will think that `setlocale` accepts invalid locale names, which in turn means that tests can't find any UTF-8 locales.

For example:
```
re/fold_grind_8.t .................................................... skipped: Couldn't find a UTF-8 locale
```

**Steps to Reproduce**
```
$ git clean -dfx && \
./Configure -des -Dusedevel &>/dev/null && grep SETLOCALE_ACCEPTS_ANY_LOCALE_NAME config.h && \
./Configure -des -Dusedevel &>/dev/null && grep SETLOCALE_ACCEPTS_ANY_LOCALE_NAME config.h
/* SETLOCALE_ACCEPTS_ANY_LOCALE_NAME:
/*#define SETLOCALE_ACCEPTS_ANY_LOCALE_NAME     / **/
/* SETLOCALE_ACCEPTS_ANY_LOCALE_NAME:
#define SETLOCALE_ACCEPTS_ANY_LOCALE_NAME       /**/
```

**Expected behavior**
The two runs of Configure should be consistent. In particular, `SETLOCALE_ACCEPTS_ANY_LOCALE_NAME` should not be defined after the second run.

**Perl configuration**

```
Summary of my perl5 (revision 5 version 39 subversion 2) configuration:
  Commit id: 292a7135a54eb726885346b222c45f30ae071632
  Platform:
    osname=linux
    osvers=6.2.0-10018-tuxedo
    archname=x86_64-linux
    uname='linux luum 6.2.0-10018-tuxedo #23 smp preempt_dynamic tue jul 18 21:29:54 utc 2023 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dusedevel -Dcc=cgcc -Accflags=-DPERL_RC_STACK'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cgcc'
    ccflags ='-DPERL_RC_STACK -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2'
    cppflags='-DPERL_RC_STACK -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='11.4.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cgcc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /usr/lib64
    libs=-lpthread -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/lib/x86_64-linux-gnu/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.35'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_RC_STACK
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under linux
  Compiled at Aug 20 2023 23:55:12
  %ENV:
    PERLBREW_BASHRC_VERSION="0.74"
    PERLBREW_HOME="/home/mauke/.perlbrew"
    PERLBREW_MANPATH="/home/mauke/perl5/perlbrew/perls/perl-5.38.0/man"
    PERLBREW_PATH="/home/mauke/perl5/perlbrew/bin:/home/mauke/perl5/perlbrew/perls/perl-5.38.0/bin"
    PERLBREW_PERL="perl-5.38.0"
    PERLBREW_ROOT="/home/mauke/perl5/perlbrew"
    PERLBREW_VERSION="0.94"
    PERLDOC="-oterm"
    PERL_UNICODE="SAL"
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.39.2/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.39.2
    /usr/local/lib/perl5/5.39.2/x86_64-linux
    /usr/local/lib/perl5/5.39.2
```

**Comment by Tux** [comment]
@khwilliamson is this your focus-area?

**Comment by khwilliamson** [comment]
Yes, but I thought this got fixed, but I don't see it

**Comment by mauke** [comment]
Nope, still broken in blead as of now.

As far as I can tell, the issue is around line 17361 of Configure. In the initial run, Configure compiles and runs a C program to detect locale features, which works. But in the second run, `d_setlocale` is already set (I think), so it uses the hardcoded defaults for whether the `C.UTF-8` locale exists (no) and whether `setlocale` is broken and accepts any locale name (yes), no matter what the previous run did.

**Comment by khwilliamson** [comment]
Fixed by #21405
